src/Symfony/Component/Serializer/InvariantViolation.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer;
+
+class InvariantViolation
+{
+    /**
+     * @var mixed
+     */
+    private $normalizedValue;
+
+    /**
+     * @var string
+     */
+    private $message;
+
+    /**
+     * @var \Throwable|null
+     */
+    private $exception;
+
+    public function __construct($normalizedValue, string $message, ?\Throwable $exception = null)
```

Comment:
I didn't see any spots where you are NOT passing the exception. So, do we need the ` = null`?

And, in all the cases you instantiate this object, it looks like throwable is NEVER null. So is the `?` needed?

Reply:
I was thinking that maybe some denormalizers would be able to create an `InvariantViolation` without catching an exception first. But that would mean denormalizers know invariant rules, while the purpose of this PR is to rely on domain objets to do that. Denormalizers should only know what kind of exceptions can be converted into `InvariantViolation` objets.

`$exception` argument is now mandatory.